### PR TITLE
Add buffering to DPRINT

### PR DESCRIPTION
**Pull request by sagarwalTT**
#14827 
#5945
#6859

This PR implements host-side and device-side buffering for DPRINT so that DPRINTs are not interleaved.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by tt-dma — APPROVED** [review]
Looks good to me, can we just add a note in the docs to show that the buffers are flushed on newline and device close? Just so users aren't surprised

File: tt_metal/impl/debug/dprint_server.cpp [hunk 1]
```diff
@@ -809,145 +891,169 @@ bool DebugPrintServerContext::PeekOneHartNonBlocking(
...
-                    // terminating zero was included in size and should be present in the buffer
-                    cptr = reinterpret_cast<const char*>(ptr);
-                    nlen = strnlen(cptr, sizeof(DebugPrintMemLayout::data));
-                    if (nlen >= 200) {
-                        stream << "STRING BUFFER OVERFLOW DETECTED" << endl;
+                {
+                    // null terminating char was included in size and should be present in the buffer
+                    const char* cptr = reinterpret_cast<const char*>(ptr);
+                    const size_t cptr_len = strnlen(cptr, sizeof(DebugPrintMemLayout::data) - 2);
+                    if (cptr_len == sizeof(DebugPrintMemLayout::data) - 2) {
```
[the remark this answers is not in the document]
Comment (sagarwalTT) [reply tt_metal/impl/debug/dprint_server.cpp:898]: This condition should only be true if the buffer doesn't end with a null terminating char. I don't think there should be a case where the device adds data that is not null terminated to the buffer, but I left this condition in just as a safety check.

**Comment by sagarwalTT** [comment]
> Looks good to me, can we just add a note in the docs to show that the buffers are flushed on newline and device close? Just so users aren't surprised

@tt-dma I've updated the doc. Can you take a look at it to see if I added it in the right place and whether I should add any more detail before I merge it in?

File: tt_metal/impl/debug/dprint_server.cpp [hunk 2]
```diff
@@ -600,6 +664,14 @@ void DebugPrintServerContext::AttachDevice(Device* device) {
         }
     }
 
+    device_intermediate_streams_force_flush_lock_.lock();
```
Comment (nhuang-tt) [inline tt_metal/impl/debug/dprint_server.cpp:667]: @sagarwalTT I started seeing seeing an assertion on main with a debug build coming from here

```
export TT_METAL_DPRINT_CORES=all
export TT_METAL_DPRINT_ETH_CORES=all
export TT_METAL_DPRINT_CHIPS=0,1
``

Comment (sagarwalTT) [reply tt_metal/impl/debug/dprint_server.cpp:667]: what is the assertion error you're encountering?

Comment (nhuang-tt) [reply tt_metal/impl/debug/dprint_server.cpp:667]: Nvm. Something is wrong with the state of my system...

**Pull request merged**